File: src/main/java/com/waridley/chatgame/game/GameObject.java
package com.waridley.chatgame.game;

import org.bson.codecs.pojo.annotations.BsonIgnore;

import java.util.List;

public abstract class GameObject {
	
	@BsonIgnore
	protected GameObject parent = null;
	
	protected List<GameObject> children;
	
	private void addChild(GameObject child) throws Exception {
		if(child.parent == null) {
			children.add(child);
			child.parent = this;
		} else {
			throw new Exception("Cannot claim child -- child already has a parent!");
		}
	}
}

File: game/src/main/java/com/waridley/chatgame/game/GameObject.kt
package com.waridley.chatgame.game

import org.bson.codecs.pojo.annotations.BsonIgnore

open class GameObject {
	
	@BsonIgnore
	protected var parent: GameObject? = null
	protected val children: MutableList<GameObject> = ArrayList()
	
	@BsonIgnore
	@Throws(Exception::class)
	private fun addChild(child: GameObject) {
		if (child.parent == null) {
			children.add(child)
			child.parent = this
		} else {
			throw Exception("Cannot claim child -- child already has a parent!")
		}
	}
}